src/dune_rules/lib.ml

```diff
@@ -1017,21 +1014,32 @@ end = struct
   end
 
   let second_step_closure ts impls =
-    let visited = ref Id.Set.empty in
-    let res = ref [] in
+    let module R = struct
```

Comment:
This code is perfectly fine, however last time we checked with @snowleopard such locally applied functors are unfortunately not optimised by the compiler. This means that every time `second_step_closure` is called the `State.Make` functor is going to be re-applied, causing additional work and allocations.

For this reason, we should manually lift this `R` module outside of the body of this function.

Reply:
Good call. Done.